### PR TITLE
Add check for a substantial annotation when loading amr counts

### PR DESCRIPTION
**Pull request by lvreynoso**
# Description

Meant to handle this error in `pipeline_run.rb`, method `db_load_amr_counts`:

```
irb(main):003:0> Sample.find(26995).pipeline_runs[0].db_load_amr_counts
Traceback (most recent call last):
        3: from (irb):3
        2: from app/models/pipeline_run.rb:552:in `db_load_amr_counts'
        1: from app/models/pipeline_run.rb:558:in `block in db_load_amr_counts'
NoMethodError (undefined method `split' for nil:NilClass)
```

It appears that there are sequences without annotations in the ARG Annot database. This pr handles that case and adds error checking for _every_ key it tries to read.

Also, too, I added more Rspec tests specifically to make sure `db_load_amr_counts()` can handle what we can throw at it.

# Tests

* Run rspec test with `docker-compose run --rm web "RAILS_ENV=test rspec spec/models/pipeline_run_spec.rb"`

### CODE REVIEW

File: app/models/pipeline_run.rb [hunk 1]
```diff
@@ -550,18 +550,23 @@ def db_load_amr_counts
...
-          genbank_accession: row["annotation"].split(";")[4],
           drug_family: row["gene_family"],
           total_reads: row["total_reads"],
           rpm: row["rpm"],
           dpm: row["dpm"],
         }
+        if row["annotation"].present?
+          if row["annotation"].split(";").length >= 5
+            amr_count_for_gene[:annotation_gene] = row["annotation"].split(";")[2]
+            amr_count_for_gene[:genbank_accession] = row["annotation"].split(";")[4]
```
Comment (davidrissato) [inline app/models/pipeline_run.rb]: suggestion: apply DRY concepts here
```
   if row["annotation"].present?
          split_annotation=row["annotation"].split(";")
          if split_annotation.length >= 5
            amr_count_for_gene[:annotation_gene] = split_annotation[2]
            amr_count_for_gene[:genbank_accession] = split_annotation[4]
```

File: spec/models/pipeline_run_spec.rb [hunk 2]
```diff
@@ -1,6 +1,31 @@
 require 'rails_helper'
 
-describe PipelineRun, type: :model do
+MALFORMED_AMR_RESULTS = [
+  ["Sample", "DB", "gene", "allele", "coverage", "depth", "diffs", "uncertainty", "divergence", "length", "maxMAF", "clusterid", "seqid", "annotation", "gene_family", "total_gene_hits", "total_coverage", "total_depth", "total_reads", "rpm", "dpm"],
+  ["", "ARGannot_r2", "Aph3''Ia_AGly", "Aph3''Ia_1217", "4.779", "1.875", "1snp777holes", "edge0.0", "2.5639999999999996", "816", "0.0", "251", "1217", "", "AGly", "1", "4.779", "1.875", "3", "0.02481738581632464", "0.0155108661352029"],
+  ["", "ARGannot_r2", "MeowZ", "", "4.779", "1.875", "1snp777holes", "edge0.0", "2.5639999999999996", "816", "0.0", "251", "1217", "", "", "1", "4.779", "1.875", "3", "0.02481738581632464", ""],
+].freeze
+MALFORMED_AMR_COUNTS = [
```
Comment (tfrcarvalho) [inline spec/models/pipeline_run_spec.rb]: For readability, you could define each line as a constant whose name identifies the type of malformation you are testing for. And then you can set: 
```
MALFORMED_AMR_COUNTS = [
	VALUES_WITH_SINGLE_QUOTES,
	...
]
```

I'm actually not sure what is wrong with the second entry...

File: spec/models/pipeline_run_spec.rb [hunk 3]
```diff
@@ -158,4 +183,68 @@
       end
     end
   end
+  context "ensure that loading amr counts works properly" do
```
Comment (tfrcarvalho) [inline spec/models/pipeline_run_spec.rb]: Do these tests trigger the behavior you were seeing with the annotation string and the ';'s?

File: spec/models/pipeline_run_spec.rb [hunk 4]
```diff
@@ -158,4 +183,68 @@
...
+        allow(Rails.logger).to receive(:error)
+      end
+      it "should log an error and simply return" do
+        pipeline_run.db_load_amr_counts()
+
+        expect(Rails.logger).to have_received(:error).with(match(/No AMR results file found for PipelineRun #/))
+        expect(pipeline_run.amr_counts).to eq([])
+      end
+    end
+    context "AMR counts file exists, but is empty (results found no amr counts)" do
```
Comment (jshoe) [inline spec/models/pipeline_run_spec.rb]: nit: could you add blank lines between the blocks for readability?

File: spec/models/pipeline_run_spec.rb (hunk 3 above)
Comment (lvreynoso) [reply spec/models/pipeline_run_spec.rb]: Nah cause I took out the regex

Comment (tfrcarvalho) [reply spec/models/pipeline_run_spec.rb]: But they should rigger that use case, to make sure it is fixed...

Comment (lvreynoso) [reply spec/models/pipeline_run_spec.rb]: Yeah it's in there!

**Comment by MarkAZhang** [comment]
LGTM! But I'll leave the final approval to someone who's familiar with why the first fix didn't work.

**Pull request merged**